PARQUET-1332: Add bloom filter for parquet

cjjnjust opened the pull request:
This is first part of bloom filter patch set, which include a bloom filter utility and also some unit tests.  
Note that this patch also includes murmur3Hash original code from Austin Appleby. The code isn't formatted as parquet-cpp format.

daedric commented on src/parquet/bloom.cc:
```diff
@@ -0,0 +1,215 @@
+    this->hashFunc = &MurmurHash3_x64_128;
+    break;
+  default:
+    throw new parquet::ParquetException("Not supported hash strategy");
```
are you sure you want to new the exception ?

cjjnjust replied:
nice catch, thanks.

daedric commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,178 @@
+  Algorithm algorithm;
+
+  // The underlying byte array for bloom filter bitset.
+  uint8_t* bitset;
```
The move ctor and operator are not deleted which could lead to a crash since the bitset is freed in the dtor.
Good practice would be to wrap it in a std::unique_ptr<uint8_t[]>

cjjnjust replied:
Good suggestion. applied.

daedric commented on src/parquet/bloom.cc:
```diff
@@ -0,0 +1,216 @@
+
+Bloom::~Bloom() {
+  if (bitset) {
+    free(bitset);
```
use delete[] bitset; it was `new`ed not `malloc`ed

cjjnjust commented:
Not sure why CI build error,  valgrind complain for "jump or move depends on uninitialised value".

daedric commented:
You don't memset to 0 your int array so it contains garbage :)

cjjnjust commented:
Hi @winningsix and @daedric 
any update on this?

cjjnjust commented:
@wesm @xhochy @daedric Since @winningsix has no bandwidth to handle this, I 'm back to take this again. This has been reviewed for long time, could you please kindly help to have a look again?

jbapple-cloudera commented on src/parquet/bloom.cc:
```diff
@@ -0,0 +1,217 @@
+  sink->Write(bitset.get(), num_bytes);
+}
+
+Bloom::~Bloom() {
```
You can put this in .h with and use " = default;"

cjjnjust replied:
done.

jbapple-cloudera commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,178 @@
+  Algorithm algorithm;
+
+  // The underlying byte array for bloom filter bitset.
+  std::unique_ptr<uint8_t[]> bitset;
```
This is almost always going to be used as a `uint32_t[]`, yes? Can we use that instead of `uint8_t[]`?

cjjnjust replied:
done.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,178 @@
+  static constexpr uint32_t SALT[8] = { 0x47b6137bU, 0x44974d91U, 0x8824ad5bU,
+      0xa2b7289dU, 0x705495c7U, 0x2df1424bU, 0x9efc4947U, 0x5c6bfb31U };
+
+  typedef void (*HashFunc)(const void *, int, uint32_t, void*);
```
yes, done

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,178 @@
+ public:
+  /// Constructor of bloom filter, if numBytes is zero, bloom filter bitset
+  /// will be created lazily and the number of bytes will be calculated through
+  /// distinct values in cache. It use murmur3_x64_128 as its default hash function
```
done.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,178 @@
+  /// Constructor of bloom filter, if numBytes is zero, bloom filter bitset
+  /// will be created lazily and the number of bytes will be calculated through
+  /// distinct values in cache. It use murmur3_x64_128 as its default hash function
+  /// and block based algorithm as default algorithm.
```
done.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,178 @@
+  // Bytes in a tiny bloom filter block.
+  static constexpr int BYTES_PER_FILTER_BLOCK = 32;
+
+  // Default seed for hash function
```
done.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,178 @@
+  // Default maximum bloom filter size (need to discuss)
+  static constexpr int DEFAULT_MAXIMUM_BLOOM_FILTER_BYTES = 16 * 1024 * 1024;
+
+  // The block based algorithm needs 8 odd SALT values to calculate eight index
```
done.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,178 @@
+  Bloom(const uint8_t* bitset, uint32_t len);
+
+  Bloom(const Bloom& orig) = delete;
+  virtual ~Bloom();
```
done.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,178 @@
+  std::unique_ptr<uint8_t[]> bitset;
+
+  // Hash function applied.
+  HashFunc hashFunc;
```
hash_strategy is used to specify which hash function to use (MurmurHash3_x86_32, MurmurHash3_x86_128, MurmurHash3_x64_128). While hashFunc is a function pointer used to point to the function we chose by hash_strategy.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,178 @@
+  void writeTo(const std::shared_ptr<OutputStream>& sink);
+
+ private:
+  // Create a new bitset for bloom filter, at least 256 bits will be create.
```
done.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,178 @@
+  // Create a new bitset for bloom filter, at least 256 bits will be create.
+  // @param numBytes number of bytes for bitset
+  void initBitset(uint32_t num_bytes);
+  void setMask(uint32_t key, uint32_t mask[8]);
```
done.

cjjnjust commented:
Thanks Jim, I have updated code accordingly. I also added an unit test for compatibility test.

Not sure why CI still failed, four build passed, two build failed.

cjjnjust commented:
Sorry, the rebase seems weird, it contains several same commits. 

I just picked up all related changes and patched to master, it overrides commit history, but I think it should be much clean to cherry pick all patches. It may need you to check previously comments by manually click "show outdate".

jbapple-cloudera commented on src/parquet/CMakeLists.txt:
```diff
@@ -50,6 +52,7 @@ install(FILES
   "${CMAKE_CURRENT_BINARY_DIR}/parquet.pc"
   DESTINATION "${CMAKE_INSTALL_LIBDIR}/pkgconfig/")
 
+ADD_PARQUET_TEST(bloom-test)
```
Is this only used in the test? What must be done to actually integrate it with the rest of the repository and get it into some binary tools or libraries that can be used by consumers of parquet-cpp build artifacts?

cjjnjust replied:
Yes, this is only expected to be used in test. I don't expect this unit test to be contained in final binary file. I think this test is used to verify correctness and compatibility before parquet-cpp releasing.

jbapple-cloudera replied:
I don't think I understand. If this file isn't used anywhere else in the project, how can it be actually used? @rdblue , maybe you could shed some light on how new features are usually integrated into the non-test parts of this project.

cjjnjust replied:
this file is used when execute "make unittest" command, it run several unit test. But the unit test code is not linked into finally parquet-cpp binary.

cjjnjust replied:
My bad, I thought it is not linked to, while I just see it is linked into final binary in compiling log.  Let me investigate how other unit test achieve this.

cjjnjust replied:
Just reviewed this again, bloom-test.cc is not linked to either libparquet.a or libparquet.so, the logs I mentioned above is linking bloom-test.o to bloom-test binary in debug/bloom-test.

jbapple-cloudera replied:
I am not comfortable that this patch is complete until it is clearly used somewhere *other* than a test. If BFs are part of the format, and this repo includes utilities for reading or writing the format, they should include this code in their compiled binaries.

cjjnjust replied:
I understand your concern while we should discussed in previous parquet sync-up meeting and agreed to implement BF utility in the test scope firstly since BFs are optional of metadata.  BFs are planed to be used when either read or write side is implemented, we should include some usages right in next steps, it that OK?

jbapple-cloudera replied:
OK. Please note as such in your commit message.

jbapple-cloudera commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,191 @@
+ public:
+  /// Constructor of bloom filter. The range of num_bytes should be within
+  /// [BYTES_PER_FILTER_BLOCK, DEFAULT_MAXIMUM_BLOOM_FILTER_BYTES], it will be
+  /// round up/down to lower/upper bound if num_bytes is out of range. It will use
```
nit: "rounded up/down"

cjjnjust replied:
Will update in next commit.

jbapple-cloudera commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,191 @@
+  /// Constructor of bloom filter. The range of num_bytes should be within
+  /// [BYTES_PER_FILTER_BLOCK, DEFAULT_MAXIMUM_BLOOM_FILTER_BYTES], it will be
+  /// round up/down to lower/upper bound if num_bytes is out of range. It will use
+  /// murmur3_x64_128 as its default hash function and the block based algorithm.
```
I'm not sure it makes sense for this class to include both an implementation and the interface of a filter. Would it make more sense to define a  pure virtual Filter interface and then derive from that and override methods?

Shall we do the same for the hash function?

cjjnjust replied:
The idea for using a function pointer is to treat murmur3 hash implementation as library functions,  such as [library ](https://docs.projects.genivi.org/ipc.common-api-dbus-runtime/2.0.4/MurmurHash3_8h.html) here.  So that one day if we want to switch to other hash in a c++ library, we only need to just assign the function pointer to it instead of change internal implementation.

jbapple-cloudera replied:
I'm sorry, but I'm not sure this addresses my concerns either about the Filter API or the Hash API. Pure abstract classes is a common idiom that more simply captures the intent that there is one API and multiple implementations.

The Derived classes, can, of course, call Murmur3 or whatever.

cjjnjust replied:
Understood your point, will update after next commit.

jbapple-cloudera commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,191 @@
+  /// function and block-based algorithm as default algorithm.
+  /// @param bitset The given bitset to construct bloom filter.
+  /// @param len Length of bitset.
+  Bloom(const uint8_t* bitset, uint32_t len);
```
This is a little tricky: If we are going to treat `bitset` as if it were an array of `uint32_t`s, we run into http://eel.is/c++draft/basic.lval#11

> If a program attempts to access the stored value of an object through a glvalue of other than one of the following types the behavior is undefined:
>  * the dynamic type of the object,
> ...
>  * a char, unsigned char, or std​::​byte type.

So we need `uint8_t` to be `char`, `unsigned char`, or `std::byte`. We'll need to check for alignment, as well, because compilers are allowed to assume that pointers to `uint32_t` have alignment that is greater than that of `uint8_t`.

Let's think about how this will get hooked into other libraries an ask @rdblue , who is more experienced in this code base, if we can put alignment requirements on the data we'll end up needing to use this with.

In an ideal world, I'd like it to be 64-byte aligned.

cjjnjust replied:
Generally types alignment is equal to types size ([Wiki](https://en.wikipedia.org/wiki/Data_structure_alignment), this may be different among compilers), so I think the last sentence you mentioned should be 64-bits aligned, right? 

Therefore, uint32_t should have an four bytes alignment generally, consider we are restricting bitset length to [32bytes, 16Mbytes], and also length is power of 2, then the uint32_t* must not contains any padding, right?

BTW, as I known, uint8_t is typedef of unsigned char, am I wrong?

jbapple-cloudera replied:
I do really mean 64-byte. That is the cache line size in most modern x86-64 CPUs. Alignment is especially useful for SIMD operations, in which misaligned access can cause a segfault.

cjjnjust replied:
OK,  Now I understood your concern.  SIMD operation should have some performance penalty when access unaligned data due to cross cache line access or cache miss (Some tests here: http://www.agner.org/optimize/blog/read.php?i=415#423), but I think it should not cause a segfault. 

How about set minimum bloom filter size to 64 bytes to avoid this?

cjjnjust replied:
I should  have misunderstanding here cause I thought set to 64 bytes will cover this.

I will ask @rdblue to share his points on this.

jbapple-cloudera replied:
Objects can have alignment that is smaller than their size:

    struct Array { char payload[11]; };
    static_assert(alignof(Array) < sizeof(Array),"");

cjjnjust replied:
I agree, the structure should have default alignment of its maximum member alignment. 

Any other thought or advice on this?

cjjnjust replied:
For 64 bytes alignment, how about we allocate  (size+63) bytes, and use std::align to do a alignment. Or use boost::alignment:aligned_alloc.

cjjnjust replied:
I got your point from your email reply. 

How about we check bitset address alignment and do copy if not?  And of course document alignment requirement and performance impact.  Can we do this firstly and consider BF client/consumer when doing read-side task?

cjjnjust replied:
Do you mean this kind segfalut? https://stackoverflow.com/questions/25596379/simd-intrinsics-segmentation-fault

cjjnjust replied:
I can do this. I don't understand the last sentence, do you mean I should add alignment check and static_assert code as well?

jbapple-cloudera replied:
Yes, add these in the test code. The return value you get from an allocatino should be aligned at at least alignof(uint32_t) and also uint8_t should be std::is_same to unsigned char.

jbapple-cloudera commented on src/parquet/bloom.cc:
```diff
@@ -0,0 +1,215 @@
+
+uint64_t Bloom::hash(const Int96 *value) {
+  uint64_t out[2];
+  (*hashFunc)(reinterpret_cast<void *>(const_cast<uint32_t*>(value->value)),
```
This `const_cast` is not necessarily standards-compliant. I expect the type signature of the hash function is a problem: it should not be modifying the data passed into it, so it should be OK with a `const` pointer.

cjjnjust replied:
yes, the const_cast is not necessarily and the signature is right also, I need to change reinterpret_cast<void*> to reinterpret_cast<const void *>

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,191 @@
+  static constexpr int BYTES_PER_FILTER_BLOCK = 32;
+
+  // Default seed for hash function, to keep cross compatibility the seed is same as
+  // value defined in parquet-mr.
```
The seed is a primer which comes from org.apache.hive.common.util.Murmur3.java. I haven't found why it use this primer, but this value is widely used in different versions of implementation.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,191 @@
+  static constexpr double DEFAULT_FPP = 0.01;
+
+  // Bloom filter data header, including number of bytes, hash strategy and algorithm.
+  static constexpr int HEADER_SIZE = 12;
```
I will remove this.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,191 @@
+  // value defined in parquet-mr.
+  static constexpr int DEFAULT_SEED = 104729;
+
+  // Default maximum bloom filter size (need to discuss)
```
Originally I think this would need to discuss with parquet community, maybe we can set to 1/8 or parquet.block.size in case of one column of long type with its distinct values ratio equal to 1 and FPP set to 0.01. What do you think?

jbapple-cloudera replied:
What would it mean to set this to 1/8? I don't understand. Doesn't this need to be an integer?

I would suggest starting the conversation on dev@ if you think it needs to be discussed by the community before committing.

cjjnjust replied:
Yes, I will start thread on this.

jbapple-cloudera commented on src/parquet/bloom.cc:
```diff
@@ -0,0 +1,215 @@
+  addElement(hash);
+}
+
+uint64_t Bloom::hash(int value) {
```
What I mean is that most of these  could be replaced by a single template.

cjjnjust replied:
OK, will do this in next commit.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,182 @@
+#ifndef PARQUET_BLOOM_H
+#define PARQUET_BLOOM_H
+
+#include <set>
```
Will remove in next commit.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,182 @@
+  // @param ndv: The number of distinct values.
+  // @param fpp: The false positive probability.
+  // @return optimal number of bits of given n and p.
+  static uint32_t optimalNumOfBits(uint32_t ndv, double fpp);
```
Will update function names in next commit.

parquet-cpp already have cpplint tool integrated, we can run make lint to check.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,182 @@
+  // Determine whether an element exist in set or not.
+  // @param hash the element to contain.
+  // @return false if value is definitely not in set, and true means PROBABLY in set.
+  bool find(uint64_t hash);
```
Agreed, will update in next commit.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,182 @@
+  ~Bloom() = default;
+
+  // Calculate optimal size according to the number of distinct values and false
+  // positive probability.
```
do you mean add a blank line between function description and param description?

jbapple-cloudera replied:
not a blank line, a line with just the comment characters and a newline.

In general, for style comments like this that you're not clear on, please check the other C++ files in this repo and copy their style.

cjjnjust replied:
will update in next commit.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -0,0 +1,165 @@
+}
+
+TEST(FPPTest, TestBloomFilter) {
+    int exist = 0;
```
will update in next commit.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -0,0 +1,165 @@
+
+TEST(FPPTest, TestBloomFilter) {
+    int exist = 0;
+    std::string str("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz");
```
will update in next commit.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    // The exist should be probably less than 1000 according default FPP 0.01.
+    ASSERT_TRUE(exist < 1000);
```
will update in next commit.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -0,0 +1,165 @@
+
+  // Deserialize bloom filter from memory
+  std::shared_ptr<InMemoryInputStream> source(new InMemoryInputStream(sink->GetBuffer()));
+  int64_t bytes_avaliable;
```
will update in next commit

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -0,0 +1,165 @@
+  int64_t bytes_avaliable;
+  uint32_t length = *(reinterpret_cast<const uint32_t *>(
+      source->Read(4, &bytes_avaliable)));
+  ASSERT_EQ(length, 1024);
```
will do in next commit.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -0,0 +1,165 @@
+TEST(FPPTest, TestBloomFilter) {
+    int exist = 0;
+    std::string str("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz");
+    std::vector<std::string> random_strings;
```
Agreed, will update in next commit.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -0,0 +1,165 @@
+      bloom->insert(bloom->hash(&byte_array));
+    }
+
+    std::shared_ptr<InMemoryOutputStream> sink;
```
Yes, I can update this.  For InMemoryOutputStream usage, seems other unit tests such as column-writer-test use in smart pointer way. So I just refer to that.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -0,0 +1,165 @@
+    int exist = 0;
+    std::string str("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz");
+    std::vector<std::string> random_strings;
+    std::unique_ptr<Bloom> bloom(new Bloom(Bloom::optimalNumOfBits(100000, 0.01)));
```
It can be, I just refer to this smart pointer usage from other unit tests.

jbapple-cloudera replied:
Please do simplify it by removing the unnecessary heap allocations (here and elsewhere). Looking at the rest of the code, I see that the stack is used almost all the time when possible.

cjjnjust replied:
Already removed in my code base, you will see this in next commit.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -0,0 +1,165 @@
+
+    // Insert 100000 elements to bloom filter and serialize to memory
+    std::random_device rd;
+    std::mt19937 generator(rd());
```
OK, will update in next commit.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -0,0 +1,165 @@
+    std::random_device rd;
+    std::mt19937 generator(rd());
+    for (int i = 0; i < 100000; i++) {
+      std::shuffle(str.begin(), str.end(), generator);
```
This just one way to generate random string, maybe the benefit is less code. Any benefit one generate random string character-by-character?

jbapple-cloudera replied:
Yes - it will be simpler, it will be less code, and it will be idiomatic.

I'd suggest pulling it out into its own simple method.

cjjnjust replied:
Not get the point, do you mean use use generate random string character-by-character?

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -0,0 +1,165 @@
+    for (int i = 0; i < 100000; i++) {
+      ByteArray byte_array1(10,
+          reinterpret_cast<const uint8_t*>(random_strings[i].c_str()));
+      ASSERT_TRUE(de_bloom->find(de_bloom->hash(&byte_array1)));
```
I didn't found other unit tests use CHECK macros yet. Do you mean DCHECK? Could you please list one code example use CHECK so that I can refer to.

jbapple-cloudera replied:
You're right - I meant `EXPECT`: https://github.com/google/googletest/blob/master/googletest/docs/primer.md

cjjnjust replied:
OK, will update to use EXPECT_EQ in next commit.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -0,0 +1,165 @@
+}//namespace test
+
+
+}// namespace parquet
```
will perform this before next commit.

cjjnjust commented on src/parquet/bloom.cc:
```diff
@@ -0,0 +1,214 @@
+  }
+}
+
+bool Bloom::contains(uint64_t hash) {
```
will remove contains and addElements in next commit.

cjjnjust commented on src/parquet/bloom.cc:
```diff
@@ -0,0 +1,214 @@
+  return out[0];
+}
+
+void Bloom::writeTo(const std::shared_ptr<OutputStream>& sink) {
```
You will see raw pointer in next commit

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,182 @@
+ public:
+  // Hash strategy available for bloom filter.
+  enum HashStrategy {
+    MURMUR3_X64_128
```
will update in next commit.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,182 @@
+  void writeTo(const std::shared_ptr<OutputStream>& sink);
+
+ private:
+  // Create a new bitset for bloom filter, the size will be at least 32 bytes.
```
will update in next commit

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,182 @@
+  // Create a new bitset for bloom filter, the size will be at least 32 bytes.
+  // @param num_bytes number of bytes for bitset. The range of num_bytes should be within
+  // [BYTES_PER_FILTER_BLOCK, DEFAULT_MAXIMUM_BLOOM_FILTER_BYTES], it will be
+  // rounded up/down to lower/upper bound if num_bytes is out of range.
```
will update in next commit

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -0,0 +1,165 @@
+    ASSERT_TRUE(exist < 1000);
+}
+
+TEST(CompatibilityTest, TestBloomFilter) {
```
will add in next commit.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -0,0 +1,165 @@
+
+    std::string testString[4] = {"hello", "parquet", "bloom", "filter"};
+	std::string data_dir(test::get_data_dir());
+    std::string bloomDataPath = data_dir + "/bloom_filter.bin";
```
Can we add doc at begin of this test?  Add README file in source subdir looks a bit weird and may be ignored easily.

cjjnjust commented:
Thanks Jim for these really great comments!

I 'd like to update some simple changes firstly, some other comments need more investigation and discussion will be updated after discussing with community.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,182 @@
+  bool contains(uint64_t hash);
+
+  // The number of bytes of bloom filter bitset.
+  uint32_t num_bytes;
```
Done.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -0,0 +1,182 @@
+  // Hash strategy used in this bloom filter.
+  HashStrategy hash_strategy;
+
+  // Algorithm applied of this bloom filter.
```
Done.

cjjnjust commented:
@jbapple-cloudera Sure, thanks.

jbapple-cloudera commented on src/parquet/bloom.cc:
```diff
@@ -0,0 +1,158 @@
+}
+
+void Bloom::WriteTo(OutputStream* sink) {
+  sink->Write(reinterpret_cast<const uint8_t*>(&num_bytes_), sizeof(uint32_t));
```
The problem here is that you don't know that num_bytes_ has size sizeof(uint32_t). Please use an explicit underlying type: https://en.cppreference.com/w/cpp/language/enum

I mentioned this in a previous comment, to which you replied "will update in next commit." I assume you fixed the part of the comment you did understand but did not quite catch my meaning about the underlying type. That happens all the time, and it's nothing to be uncomfortable about. I strongly suggest asking for clarification if you get a comment from a code reviewer and you do not understand what it means.

cjjnjust replied:
My bad, I did miss this. I assume you are talking about add underlying type for HashStratey and Algorithm enum.

jbapple-cloudera replied:
yes

jbapple-cloudera commented on src/parquet/bloom-test.cc:
```diff
@@ -0,0 +1,187 @@
+  EXPECT_TRUE(exist < total_count * fpp);
+}
+
+// The CompatibilityTest is use to test cross compatibility with parquet-mr, it read
```
nit: "it reads"

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -32,134 +32,156 @@ namespace test {
+  bloom.WriteTo(&sink);
+
+  // Deserialize bloom filter from memory
+  InMemoryInputStream source(sink.GetBuffer());
```
Can we only perform serializing and de-serializing test only in InsertAndFindTest? And rename to BasicFuntionalityTest?

jbapple-cloudera replied:
Yes

cjjnjust commented on src/parquet/murmur3.h:
```diff
@@ -15,42 +15,50 @@
-//-----------------------------------------------------------------------------
-
-void MurmurHash3_x86_32(const void *key, int len, uint32_t seed, void *out);
+  // Default seed for hash which comes from Murmur3 implementation in Hive
```
Hive imports Murmur3 from Orc, while I still can not find how they choose this primer. Any suggestion?

jbapple-cloudera replied:
The commit that added it to Orc - what does that commit message say?

cjjnjust replied:
The original commit should be HIVE-9188, I checked code review comments(https://reviews.apache.org/r/29296/) and commit message, none of them mentioned about the seed, there is also no comment on murmur3.java.

cjjnjust replied:
I agree to make it as a parameter and it is already a parameter, do I misunderstand something? Per my understanding we are discussing how to select a default seed. I am OK to generate a new one randomly as you point out.

jbapple-cloudera replied:
"It" means the seed. You could make the seed part of the serialization of the filter and allow the user to specify the hash function and seed at the time the filter is created.

cjjnjust replied:
Can we use a new random value as your suggested and skip Serde for seed? Adding Serde logic for hash seed will need to change parquet-format, and update parquet-mr as well.  In the first stage, how about keep it simple and stupid?

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -18,174 +18,140 @@
-
-  // Default maximum bloom filter size (need to discuss)
+  // Default maximum Bloom filter size (need to discuss)
   static constexpr int DEFAULT_MAXIMUM_BLOOM_FILTER_BYTES = 16 * 1024 * 1024;
```
OK. Can we leave it is for upper bound check and update comment only?

jbapple-cloudera replied:
It's pretty low for a check. Maybe increase the size to something really enormous like 1TB and rename it MAX_SANE_FILTER_SIZE?

cjjnjust replied:
Since old lower boundary depends on constant BYTES_PER_FILTER_BLOCK which will be move out of Bloom class for algorithm abstract, So I 'm going to add two macros named MINIMUM_BLOOM_FILTER_BYTES and MAXIMUM_BLOOM_FILTER_BYTES to replace old boundaries. MINIMUM_BLOOM_FILTER_BYTES would be 64 bytes as same as cache alignment, and MAXIMUM_BLOOM_FILTER_BYTES would be 256MB (HADOOP default block size), how about this?

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -18,174 +18,140 @@
+// implementation sets 8 bits in each tiny Bloom filter. Tiny Bloom filter are 32
 // bytes to take advantage of 32-bytes SIMD instruction.
 
 class Bloom final {
```
I can add abstract for Algorithm too.

Do you suggest to respond to review (address comments) firstly and push a new commit to cover all comments?

jbapple-cloudera replied:
You state "the bloom algorithm abstract class is specific to Bloom filter, create a separated class seems too complex". I have to disagree. The idea of an interface to a filter, with methods like InsertHash and FindHash, seems not so complex. Furthermore, many possible implementation strategies exist for filters, including this block split hash, traditional bloom filters, cuckoo filters, quotient filters, and many more, and all of them support InsertHash and FindHash.

cjjnjust replied:
OK, I will do this abstract in next commit.

cjjnjust replied:
@jbapple-cloudera , I have updated code to use Bloom filter algorithm abstract class.  I didn't reply nits comments but I do update code for them since I think it would be too many noise for you.

cjjnjust commented on src/parquet/bloom.cc:
```diff
@@ -92,9 +91,9 @@ void Bloom::setMask(uint32_t key, uint32_t mask[8]) {
+uint32_t Bloom::OptimalNumOfBits(uint32_t ndv, double fpp) {
   DCHECK(fpp > 0.0 && fpp < 1.0);
-  const double M = -8.0 * ndv / log(1 - pow(fpp, 1.0/8));
+  const double M = -8.0 * ndv / log(1 - pow(fpp, 1.0 / 8));
```
Do you mean to use lower case "m"?

jbapple-cloudera replied:
Yes

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -32,134 +32,156 @@ namespace test {
+  std::seed_seq seed(uuid_seed.begin(), uuid_seed.end());
+  std::mt19937 generator(seed);
+
+  // Insert 100000 elements to bloom filter and serialize to memory
```
Done.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -18,174 +18,140 @@
+                                       0x9efc4947U, 0x5c6bfb31U};
 
-  typedef void (*HashFunc)(const void *key, int length, uint32_t seed, void* out);
+  typedef void (*HashFunc)(const void* key, int length, uint32_t seed, void* out);
```
Deleted.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -18,174 +18,140 @@
+  ///
+  /// @param hash the element to contain.
+  /// @return false if value is definitely not in set, and true means PROBABLY in set.
+  bool FindHash(uint64_t hash);
```
Done.

cjjnjust commented on src/parquet/bloom.cc:
```diff
@@ -59,26 +58,26 @@ void Bloom::initBitset(uint32_t num_bytes) {
 
-  this->bitset.reset(new uint32_t[num_bytes/4]);
-  memset(this->bitset.get(), 0, num_bytes);
+  this->bitset_.reset(new uint32_t[num_bytes / 4]);
```
Done

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -18,174 +18,140 @@
+  /// @param ndv The number of distinct values.
+  /// @param fpp The false positive probability.
+  /// @return optimal number of bits of given n and p.
+  static uint32_t OptimalNumOfBits(uint32_t ndv, double fpp);
```
Done.

cjjnjust commented on src/parquet/bloom.cc:
```diff
@@ -59,26 +58,26 @@ void Bloom::initBitset(uint32_t num_bytes) {
+    : num_bytes_(num_bytes),
+      hash_strategy_(HashStrategy::MURMUR3_X64_128),
+      algorithm_(Algorithm::BLOCK) {
+  this->bitset_.reset(new uint32_t[num_bytes / 4]);
```
Done.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -32,134 +32,156 @@ namespace test {
+    ByteArray tmp(static_cast<uint32_t>(testString[i].length()),
+                  reinterpret_cast<const uint8_t*>(testString[i].c_str()));
+    EXPECT_TRUE(bloom_filter.FindHash(bloom_filter.Hash<const ByteArray*>(&tmp)));
+  }
```
Done.

cjjnjust commented on src/parquet/bloom.cc:
```diff
@@ -59,26 +58,26 @@ void Bloom::initBitset(uint32_t num_bytes) {
 }
 
-void Bloom::setMask(uint32_t key, uint32_t mask[8]) {
+void Bloom::SetMask(uint32_t key, uint32_t mask[8]) {
```
Done.

cjjnjust commented on src/parquet/bloom.h:
```diff
@@ -18,174 +18,140 @@
-  /// bloom filter from parquet file. It use murmur3_x64_128 as its default hash
+  /// Construct the Bloom filter with given bit set, it is used when reconstruct
+  /// Bloom filter from parquet file. It use murmur3_x64_128 as its default hash
   /// function and block-based algorithm as default algorithm.
```
Done.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -32,134 +32,156 @@ namespace test {
+// whether the values inserted before could be filtered or not.
+
+// The Bloom filter binary is generated by three steps in from Parquet-mr.
+// Step 1: Construct a Bloom filter with 1024 entries.
```
reworded to with 1024 bytes bitset.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -32,134 +32,156 @@ namespace test {
+  double fpp = 0.01;
+
+  // Character set used to generate random string
+  std::string charset("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz");
```
Done.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -32,134 +32,156 @@ namespace test {
+
+  std::string testString[4] = {"hello", "parquet", "bloom", "filter"};
+  std::string data_dir(test::get_data_dir());
+  std::string bloomDataPath = data_dir + "/bloom_filter.bin";
```
Done.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -32,134 +32,156 @@ namespace test {
+  Bloom de_bloom(bitset, length);
   for (int i = 0; i < 10; i++) {
-    ASSERT_TRUE(de_bloom->find(bloom->hash(i)));
+    EXPECT_TRUE(de_bloom.FindHash(de_bloom.Hash(i)));
```
Yes.

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -32,134 +32,156 @@ namespace test {
+  InMemoryInputStream source(sink.GetBuffer());
+  int64_t bytes_available;
+  uint32_t length =
+      *(reinterpret_cast<const uint32_t*>(source.Read(4, &bytes_available)));
```
Done

cjjnjust commented on src/parquet/bloom-test.cc:
```diff
@@ -32,134 +32,156 @@ namespace test {
-  MurmurHash3_x64_128(bitset, 8, Bloom::DEFAULT_SEED, &result);
-  ASSERT_EQ(result, -3850979349427597861l);
+  MurmurHash3 murmur3;
+  murmur3.Hash_x64_128(bitset, 8, MurmurHash3::DEFAULT_SEED, &result);
```
Done.

cjjnjust commented on src/parquet/murmur3.h:
```diff
@@ -15,42 +15,50 @@
 
-void MurmurHash3_x86_128(const void *key, int len, uint32_t seed, void *out);
+  MurmurHash3(const MurmurHash3&) = delete;
+  ~MurmurHash3() = default;
```
Yes.

cjjnjust commented on src/parquet/murmur3.h:
```diff
@@ -15,42 +15,50 @@
+  static constexpr int DEFAULT_SEED = 104729;
 
-void MurmurHash3_x86_128(const void *key, int len, uint32_t seed, void *out);
+  MurmurHash3(const MurmurHash3&) = delete;
```
No harm to have it, added it.

cjjnjust commented on src/parquet/murmur3.cc:
```diff
@@ -330,3 +404,48 @@ void MurmurHash3_x64_128(const void * key, const int len,
+
+uint64_t MurmurHash3::Hash(int64_t value) {
+  uint64_t out[2];
+  Hash_x64_128(reinterpret_cast<void*>(&value), sizeof(int), seed_, out);
```
Done.

cjjnjust commented on src/parquet/murmur3.h:
```diff
@@ -15,42 +15,50 @@
+  void Hash_x64_128(const void* key, int len, uint32_t seed, void* out);
 
-#endif // !defined(_MSC_VER)
+  uint64_t Hash(int32_t value);
```
Done.

cjjnjust commented on src/parquet/hasher.h:
```diff
@@ -0,0 +1,75 @@
+  ///
+  /// @param value the value to hash.
+  /// @return hash result.
+  virtual uint64_t Hash(int32_t value) = 0;
```
Done.

cjjnjust commented on src/parquet/murmur3.h:
```diff
@@ -15,42 +15,50 @@
 
-#include <stdint.h>
+  /// Original murmur3 hash implementation
+  void Hash_x86_32(const void* key, int len, uint32_t seed, void* out);
```
Done.

jbapple-cloudera commented on src/parquet/murmur3.cc:
```diff
@@ -0,0 +1,453 @@
+  *reinterpret_cast<uint32_t*>(out) = h1;
+}
+
+void Hash_x86_128(const void* key, const int len, uint32_t seed, void* out) {
```
As I requested last time, please change the last parameter to prevent passing in the wrong type.

jbapple-cloudera commented on src/parquet/murmur3.h:
```diff
@@ -0,0 +1,59 @@
+  uint64_t Hash(const FLBA* val, uint32_t len) const override;
+
+  // Default seed for hash which comes from the Murmur3 implementation in Hive
+  static constexpr int DEFAULT_SEED = 104729;
```
As you suggested, I think it's fine to generate this randomly right now and put that random value in this class.

jbapple-cloudera replied:
Please make this private

jbapple-cloudera commented on src/parquet/murmur3.h:
```diff
@@ -0,0 +1,59 @@
+	  this->seed_ = other.seed_;
+  }
+
+  uint32_t seed_;
```
Please make this private.

jbapple-cloudera commented on src/parquet/murmur3.h:
```diff
@@ -0,0 +1,59 @@
+  // Default seed for hash which comes from the Murmur3 implementation in Hive
+  static constexpr int DEFAULT_SEED = 104729;
+
+  MurmurHash3(const MurmurHash3& other) {
```
This is the same as just leaving this code out, giving it the default semantics, yes?

jbapple-cloudera commented on src/parquet/bloom_filter-test.cc:
```diff
@@ -0,0 +1,228 @@
+  InMemoryOutputStream sink;
+  bloom_filter.WriteTo(&sink);
+
+  // Deserialize Bloom filter from memory
```
And that's fine, but now that you are deserializing twice, please make a helper function.

cjjnjust commented on src/parquet/bloom_filter.cc:
```diff
@@ -0,0 +1,164 @@
+    throw parquet::ParquetException("Given length of bitset is illegal");
+  }
+
+  if ((reinterpret_cast<std::uintptr_t>(bitset) & (X64_CACHE_ALIGNMENT - 1)) == 0) {
```
Yes, it looks a bit weird. How about to make bitset_ to std::shared_ptr and accept std::shared_ptr here?

jbapple-cloudera replied:
I don't think that's the right choice here. shared_ptr very often is covering up for other design flaws.

The first step here is to understand what interface the callers need. I asked about this in the last review, but did not receive a reply.

Also, very importantly: this block still leaks memory by providing the caller no information about whether it can free the block it passed it.

cjjnjust replied:
I haven't implemented both read/write side for parquet-cpp, while in parquet-mr read side there is  one place to call the BloomFilter(uint32_t* bitset, uint32_t len) to construct a Bloom filter from the ColumnChunkMetaData.

cjjnjust commented on src/parquet/bloom_filter.cc:
```diff
@@ -0,0 +1,164 @@
+	  num_bytes = MAXIMUM_BLOOM_FILTER_BYTES;
+  }
+
+  ::arrow::Status status = pool_->Allocate(num_bytes,
```
MemoryPool::Allocate interface is defined to allocate 64 bytes aligned buffer, the default_memory_pool uses AlocateAligned to allocate the aligned buffer, it returns Status::OK() if allocates successfully.  We have a check of return status here, so it should not happen undefined behavior, right?

cjjnjust commented on src/parquet/bloom_filter_algorithm.h:
```diff
@@ -0,0 +1,87 @@
+namespace parquet {
+
+// Abstract class for Bloom filter algorithm
+class BloomFilterAlgorithm {
```
Sorry, I don't quite understand this.  Why do we need to implement data structure polymorphism other than simple algorithm polymorphsim?  From Bloom filters perspective, what it needs is testBits and setBits, abstract these two interfaces should open maximum capability to the implementer. We may not know how the data structure be designed for other algorithms.

jbapple-cloudera replied:
Abstracting at the data structure level provides much more flexibility.

Consider how you do things now: InitBitset is defined in the BloomFilter class, but that might be different for different algorithms. The way it is written now actually constrains the data structure more tightly.

cjjnjust replied:
OK. Do you mean we should move bitset and its intializer to BloomFilterAlgorithm class and then put specific initializer to implementer such as BlockBloomFilterAlgorithm?

cjjnjust replied:
I moved bitset and related operations to algorithm class. Since the bitset we get from inputstream maybe not 64-byte aligned and no way to cast it to uint32_t pointer, so I have to use memcpy here.

cjjnjust commented on src/parquet/bloom_filter.cc:
```diff
@@ -0,0 +1,164 @@
+  }
+
+  if ((reinterpret_cast<std::uintptr_t>(bitset) & (X64_CACHE_ALIGNMENT - 1)) == 0) {
+    this->bitset_ = const_cast<uint32_t*>(bitset);
```
Can we remove const qualifier for input parameter? If not, how about use another member variable to store the external bitset and do copy on write when inserting? I'm not sure whether it is too complex.

jbapple-cloudera replied:
I think the most important information we need is about how the callers will use this. Can you look into that, please?

cjjnjust replied:
When a predicate pushed down to parquet, it needs a deserialization APIs to deserialize Bloom filter from column chunk metadata, and then evaluate the value in predicate expression with Bloom filter. That 's the common usage as I known.

cjjnjust commented on src/parquet/bloom_filter-test.cc:
```diff
@@ -0,0 +1,228 @@
+
+  // The uuid_seed was generated by "uuidgen -r"
+  const std::string uuid_seed = "8de406aa-fb59-4195-a81c-5152af26433f";
+  std::seed_seq seed(uuid_seed.begin(), uuid_seed.end());
```
This can't be const.

cjjnjust commented on src/parquet/bloom_filter-test.cc:
```diff
@@ -0,0 +1,228 @@
+  std::unique_ptr<uint8_t[]> bitset;
+
+  const std::string test_string[4] = {"hello", "parquet", "bloom", "filter"};
+  std::string data_dir(test::get_data_dir());
```
test::get_data_dir() return const char*, so we need an extra std::string here to join the filename.

cjjnjust commented on src/parquet/util/test-common.h:
```diff
@@ -101,8 +102,8 @@ static inline vector<bool> flip_coins_seed(int n, double p, uint32_t seed) {
 static inline vector<bool> flip_coins(int n, double p) {
-  std::random_device rd;
-  std::mt19937 gen(rd());
+  uint64_t seed = std::chrono::high_resolution_clock::now().time_since_epoch().count();
```
this is used to fix CI build error.  It is strange since the master should already contains this code.

jbapple-cloudera replied:
Commits should be focused on one thing. Please move this to a separate pull request.

cjjnjust replied:
Yes, this is just a bit strange since the change is not mine~  I will look into this, maybe another rebase is needed.

cjjnjust commented on src/parquet/bloom_filter-test.cc:
```diff
@@ -0,0 +1,228 @@
+  BloomFilter bloom_filter(reinterpret_cast<const uint32_t*>(bitset.get()), length);
+
+  for (int i = 0; i < 4; i++) {
+    ByteArray tmp(static_cast<uint32_t>(test_string[i].length()),
```
String::length() return long, and ByteArray needs uint32_t, without static_cast<uint32_t>, the CI build fail to compile.

error: conversion to ‘uint32_t {aka unsigned int}’ from ‘std::basic_string<char>::size_type {aka long unsigned int}’ may alter its value [-Werror=conversion]

jbapple-cloudera commented on src/parquet/bloom_filter-test.cc:
```diff
@@ -0,0 +1,228 @@
+  std::vector<std::string> members;
+  BloomFilter bloom_filter(BloomFilter::OptimalNumOfBits(total_count, fpp));
+
+  // Insert elements into the Bloom filter and serialize to memory
```
This comment was not addressed.

jbapple-cloudera commented on src/parquet/bloom_filter-test.cc:
```diff
@@ -0,0 +1,228 @@
+  for (int i = 0; i < 4; i++) {
+    ByteArray tmp(static_cast<uint32_t>(test_string[i].length()),
+                  reinterpret_cast<const uint8_t*>(test_string[i].c_str()));
+    EXPECT_TRUE(bloom_filter.FindHash(bloom_filter.Hash<const ByteArray*>(&tmp)));
```
This comment was not addressed.

jbapple-cloudera commented on src/parquet/bloom_filter-test.cc:
```diff
@@ -0,0 +1,261 @@
+// Also it is used to test whether OptimalNumOfBits returns value between
+// [MINIMUM_BLOOM_FILTER_SIZE, MAXIMUM_BLOOM_FILTER_SIZE].
+TEST(OptimalValueTest, TestBloomFilter) {
+  EXPECT_GT(BloomFilter::OptimalNumOfBits(256, 0.01), UINT32_C(2048));
```
What is the rationale for using EXPECT_GT, EXPECT_GE, or EXPECT_LE? My suggestion in the last review was not to use inequalities rather than euqalities, but only to not write code like EXPECT_TRUE(p < q).

jbapple-cloudera commented on src/parquet/bloom_filter.h:
```diff
@@ -0,0 +1,165 @@
+  /// @param num_bytes  The number of bytes of given bitset.
+  BloomFilter(const uint32_t* bitset, uint32_t num_bytes);
+
+  BloomFilter(const BloomFilter& orig) = delete;
```
A move constructor might be needed for the static deserialization method I mention above.

cjjnjust replied:
We can return BloomFilter pointer in Deserialization.

jbapple-cloudera replied:
While that is technically possible, that's a somewhat outdated C++ idiom.

jbapple-cloudera commented on src/parquet/bloom_filter.h:
```diff
@@ -15,86 +15,84 @@
-                                       0xa2b7289dU, 0x705495c7U, 0x2df1424bU,
-                                       0x9efc4947U, 0x5c6bfb31U};
+  // L1 cache alignment of x86_64 architecture.
+  static constexpr uint32_t X64_CACHE_ALIGNMENT = 64;
```
This comment was not addressed.

cjjnjust replied:
I removed this macro in the new commit and defined the minimum Bloom filter size to 32. The reason why choose 64 bytes is just want to use cache line as minimum Bloom filter size.

jbapple-cloudera commented on src/parquet/murmur3.cc:
```diff
@@ -411,42 +192,34 @@ void Hash_x64_128(const void* key, const int len, const uint32_t seed,
 uint64_t HashHelper(T value, uint32_t seed) {
   HashOutput128 output;
-  Hash_x64_128(reinterpret_cast<void*>(&value), sizeof(T), seed, &output);
+  Hash_x64_128(reinterpret_cast<void*>(&value), sizeof(T), seed, output.val);
```
I'd suggest removing the special type and declaring `uint64_t output[2];` here and below.

cjjnjust commented on src/parquet/bloom_filter-test.cc:
```diff
@@ -35,14 +35,45 @@ TEST(Murmur3Test, TestBloomFilter) {
+  EXPECT_THROW(new BloomFilter(bitset2.get(), 1023), ParquetException);
+}
+
+void DeserializeBloomFilter(InputStream* input, BloomFilter** bloom_filter) {
```
The google test framework can't compile if the there is EXPECT_ macro in function which is not a test.

cjjnjust commented on src/parquet/bloom_filter-test.cc:
```diff
@@ -35,14 +35,45 @@ TEST(Murmur3Test, TestBloomFilter) {
+  int64_t bytes_available;
+
+  uint32_t len = *(
+      reinterpret_cast<const uint32_t*>(input->Read(sizeof(uint32_t), &bytes_available)));
```
OK, will change to use memcpy here.

So there is only the memcpy way to convert from uint8_t pointer to uint32_t pointer, given most of file read APIs will return const uint8_t*, is that means we can't avoid memcpy here? should we still use uint32_t pointer?

xhochy commented on src/parquet/bloom_filter_algorithm.cc:
```diff
@@ -0,0 +1,102 @@
+
+  num_bytes_ = num_bytes;
+
+  ::arrow::Status status =
```
Instead of calling raw `Allocate` use https://github.com/apache/arrow/blob/c6e33d89d10d9aa324cbd5b6776cadbb367970a7/cpp/src/arrow/buffer.h#L232 to get a Buffer. This will then be automatically be deallocated on destruction of the reference.

cjjnjust replied:
Thanks, I will use it. Just one thing needs to be consider, converting from uint8_t* to uint32_t* with reinterpret_cast may lead to undefined behavior.

Hi @jbapple-cloudera , looks like reinterpret_cast from uint8_t* to other bigger size type pointer is widely used in this project and also arrow, such as TypedBufferBuilder. Is the undefined behavior not present when the pointer's address and the buffer size is a multiply of requirement?

jbapple-cloudera replied:
https://github.com/apache/parquet-cpp/pull/425 removed `-fno-strict-aliasing` from the build flags. This makes many common calls to `reinterpret_cast` invalid. If you see lots of other use, I'd suggest starting a thread on dev@ on the subject.

xhochy commented on src/parquet/bloom_filter_algorithm.cc:
```diff
@@ -0,0 +1,102 @@
+
+  ::arrow::Status status =
+      pool_->Allocate(num_bytes, reinterpret_cast<uint8_t**>(&bitset32_));
+  if (!status.ok()) {
```
Use the `PARQUET_THROW_NOT_OK` macro to check the Status value

cjjnjust replied:
Yes, thanks.

xhochy commented on src/parquet/bloom_filter_algorithm.cc:
```diff
@@ -0,0 +1,102 @@
+BlockBasedAlgorithm::BlockBasedAlgorithm(const uint8_t* bitset, uint32_t num_bytes)
+    : BlockBasedAlgorithm(num_bytes) {
+  if (!bitset) {
+    throw ParquetException("Given bitste is NULL");
```
Type

xhochy commented on src/parquet/bloom_filter.cc:
```diff
@@ -0,0 +1,168 @@
+      this->hasher_.reset(new MurmurHash3());
+      break;
+    default:
+      throw parquet::ParquetException("Unsupported hash strategy.");
```
While you're in the `parquet` namespace, there is no need to prefix this with `parquet::`

cjjnjust replied:
ok. Thanks.

jbapple-cloudera commented on src/parquet/bloom_filter_algorithm.cc:
```diff
@@ -16,11 +16,40 @@
 namespace parquet {
 constexpr uint32_t BlockBasedAlgorithm::SALT[BITS_SET_PER_BLOCK];
 
+BlockBasedAlgorithm::BlockBasedAlgorithm(uint32_t num_bytes)
```
This is getting complex, which is why I suggest, as I mentioned a few times before, having two classes:

`Filter`, which is has pure abstract virtual methods, and `BlockSplitBloomFilter`, but not `BloomFilterAlgorithm` class.

```
struct Filter {
  virtual void InsertHash(uint64_t) = 0;
  virtual void FindHash(uint64_t) const = 0;
  // Hash methods, destructor, all pure virtual
};

struct BlockSplitBloomFilter : public Filter {
  // Constructor
  // Overrides
};
```

I suggest you remove the `BloomFilterAlgorithm` class.

cjjnjust replied:
Agree with you, that looks like simple. Just updated code, please help to have a look.

jbapple-cloudera commented on src/parquet/bloom_filter_algorithm.h:
```diff
@@ -69,15 +73,46 @@ class BlockBasedAlgorithm : public BloomFilterAlgorithm {
+  explicit BlockBasedAlgorithm(uint32_t num_bytes);
+
+  /// The constructor of the block-based algorithm. It copies the bitset as underlying
+  /// bitset because the given bitset may not satisfy the 64-byte alignment requirement
```
The alignment requirement is not 64-bytes, it is either 4 bytes (as the code is now) or 32 bytes (AVX2).

cjjnjust replied:
Did we discuss the alignment should be 64 bytes since the SIMD maybe trigger segfault if it access cross cache line data?

jbapple-cloudera commented on src/parquet/bloom_filter_algorithm.cc:
```diff
@@ -16,11 +16,40 @@
+  num_bytes_ = num_bytes;
+
+  ::arrow::Status status =
+      pool_->Allocate(num_bytes, reinterpret_cast<uint8_t**>(&bitset32_));
```
This comment from last time was not addressed, I think:

Incorrect alignment can produce incorrect compilation even without SIMD. If we treat the result like a `uint32_t *` and it is not 4-byte aligned, then the compilation can produce surprising and incorrect code.

The documentation says that Allocate returns 8-byte aligned results. Please add a check that it is at least `alignof(decltype(bitset_[0]))`.

jbapple-cloudera commented on src/parquet/bloom_filter.cc:
```diff
@@ -94,30 +71,52 @@ BloomFilter::BloomFilter(const uint32_t *bitset, uint32_t num_bytes)
+  int64_t bytes_available;
+
+  uint32_t len;
+  memcpy(&len, input->Read(sizeof(uint32_t), &bytes_available), sizeof(uint32_t));
```
Can `Read` return a `nullptr`? If so, you'll want to checkthat the result is non-null before calling `memcpy`.

cjjnjust commented on src/parquet/bloom_filter.h:
```diff
@@ -45,25 +45,39 @@ class BloomFilter final {
+  enum class HashStrategy : uint32_t { MURMUR3_X64_128 = 0 };
+
+  // Bloom filter algorithm.
+  enum class Algorithm : uint32_t { BLOCK = 0 };
```
It was used in deserialize function. It can be move back to private since I will remove algorithm and hash check in deserialize function and let constructor to check them.

jbapple-cloudera commented on src/parquet/bloom_filter.h:
```diff
@@ -77,51 +42,209 @@ class BloomFilter final {
 
-  BloomFilter(const BloomFilter& orig) = delete;
-
   /// Calculate optimal size according to the number of distinct values and false
```
This is also specific to BlockSplitBF

jbapple-cloudera commented on src/parquet/bloom_filter.h:
```diff
@@ -77,51 +42,209 @@ class BloomFilter final {
+  BlockSplitBloomFilter(const uint8_t *bitset, uint32_t num_bytes);
+
+  // Bytes in a tiny Bloom filter block.
+  static constexpr int BYTES_PER_FILTER_BLOCK = 32;
```
Why is this public? same with BITS_SET_PER_BLOCK, BlockMask, and SALT.

jbapple-cloudera commented on src/parquet/bloom_filter.h:
```diff
@@ -77,51 +42,209 @@ class BloomFilter final {
+  ///
+  /// @param hash the hash used to calculate bit index to test
+  /// @return true
+  bool TestBits(uint64_t hash) const;
```
Why have TestBits and SetBits? Why not just put the bodies of those functions in FindHash and InsertHash?

jbapple-cloudera commented on src/parquet/bloom_filter.h:
```diff
@@ -77,51 +42,209 @@ class BloomFilter final {
+  ::arrow::MemoryPool *pool_;
+
+  // The underlying buffer of bitset.
+  std::shared_ptr<Buffer> data_;
```
Since you wisely pointed out that we can get the buffer to be 64-byte aligned, you can make this a `std::unique_ptr<uint32_t>`. Sorry for the churn!

cjjnjust replied:
AllocateBuffer needs shared_ptr as input param and looks like no easy way to convert shared_ptr to unique_ptr.

jbapple-cloudera replied:
Can you help me remember: why not use `MemoryPool::Allocate` then?

cjjnjust replied:
@xhochy suggested to use AllocateBuffer so that we don't have to free the buffer manually.

jbapple-cloudera replied:
If you use `MemoryPool::Allocate` and `unique_ptr` with a [`Deleter`](https://en.cppreference.com/w/cpp/memory/unique_ptr), you don't need a `shared_ptr`. Generally, you want to avoid `shared_ptr` when `unique_ptr` will suffice.

wesm replied:
Nearly all APIs related to memory allocation in Arrow use `shared_ptr<Buffer>` (because memory / buffer sharing is fundamental to how the project works)

jbapple-cloudera replied:
Is that the right choice here - what other references do you expect there will be to this buffer other than in a single Bloom filter?

wesm replied:
Reimplementing `arrow::AllocateBuffer` and adding a custom deleter for the sake of using `unique_ptr` seems a bit elaborate to me. It would be more worth your time to add an AllocateBuffer API that returns `unique_ptr`. I opened https://issues.apache.org/jira/browse/ARROW-2998

cjjnjust replied:
Hi @jbapple-cloudera, I tried to use custom deleter but looks like no easy way to implement it.  This is because MemoryPool needs two arguments, pointer and size, to free a buffer. While a custom Deleter needs only one pointer argument. I also tried to use following way:
```cpp
  void FreeBitset(uint8_t* data, uint32_t num_bytes) {                                                                                                    
      if (data) {                                                                                                                                         
          ::arrow::MemoryPool *pool = ::arrow::default_memory_pool();                                                                                     
          pool->Free(data, num_bytes);                                                                                                                    
      }                                                                                                                                                   
  }              
  auto deleter= [&](uint8_t* p) { FreeBitset(p, ???); }; 
  std::unique_ptr<uint8_t[], decltype(deleter)> data_;

```
But don't know how to pass in the num_bytes argument. 

I saw you have assigned the ARROW-2998 to yourself also....

jbapple-cloudera commented on src/parquet/bloom_filter.cc:
```diff
@@ -52,54 +47,50 @@ BloomFilter::BloomFilter(uint32_t num_bytes)
-    default:
-      throw parquet::ParquetException("Unsupported Bloom filter algorithm");
-  }
+  num_bytes_ = num_bytes;
```
Move this to initializer section on lines 58-60.

jbapple-cloudera commented on src/parquet/bloom_filter.h:
```diff
@@ -77,51 +42,209 @@ class BloomFilter final {
+  explicit BlockSplitBloomFilter(uint32_t num_bytes);
+
+  /// The constructor of BlockSplitBloomFilter. It copies the bitset as underlying
+  /// bitset because the given bitset may not satisfy the 64-byte alignment requirement
```
32-byte. SIMD reads that are 32-byte aligned do not cross cache line boundaries, because if and address P is divisible by 32, then no number in [P, P+32) is divisible by 64 other than possibly P itself.

jbapple-cloudera commented on src/parquet/bloom_filter.h:
```diff
@@ -77,51 +42,209 @@ class BloomFilter final {
+  /// [MINIMUM_BLOOM_FILTER_BYTES, MAXIMUM_BLOOM_FILTER_BYTES], it will be
+  /// rounded up/down to lower/upper bound if num_bytes is out of range and also
+  /// will be rounded up to a power of 2. It will use murmur3_x64_128 as its
+  /// default
```
nit: odd line break here

jbapple-cloudera commented on src/parquet/bloom_filter.h:
```diff
@@ -77,51 +42,209 @@ class BloomFilter final {
+  /// rounded up/down to lower/upper bound if num_bytes is out of range and also
+  /// will be rounded up to a power of 2. It will use murmur3_x64_128 as its
+  /// default
+  /// hash function and the block-based algorithm.
```
Don't need to say the default algorithm, only the hash function. Here and below.

jbapple-cloudera commented on src/parquet/bloom_filter.cc:
```diff
@@ -52,54 +47,50 @@ BloomFilter::BloomFilter(uint32_t num_bytes)
-BloomFilter* BloomFilter::Deserialize(InputStream* input) {
+BloomFilter* BlockSplitBloomFilter::Deserialize(InputStream* input) {
   int64_t bytes_available;
+  const uint8_t* read_buffer = NULL;
```
Delay this definition until the first assignment on line 82.

jbapple-cloudera commented on src/parquet/bloom_filter.cc:
```diff
@@ -52,54 +47,50 @@ BloomFilter::BloomFilter(uint32_t num_bytes)
+    throw ParquetException("Failed to deserialize from input stream");
+  }
+  memcpy(&hash, read_buffer, sizeof(uint32_t));
   if (static_cast<uint32_t>(bytes_available) != sizeof(uint32_t)) {
```
This is duplicated with the code on line 90. Same for line 107 and 103, respectively.

jbapple-cloudera commented on src/parquet/bloom_filter.cc:
```diff
@@ -108,61 +99,81 @@ BloomFilter* BloomFilter::Deserialize(InputStream* input) {
   }
 
-  return new BloomFilter(input->Read(len, &bytes_available), len);
+  return new BlockSplitBloomFilter(input->Read(len, &bytes_available), len);
```
Please return a value, not a `new`-allocated pointer.

wesm commented:
I will give this a review when I can. It would be valuable to have bloom filters available in Apache Arrow also, related to our mailing list discussion happening right now https://lists.apache.org/thread.html/4bc135b4e933b959602df48bc3d5978ab7a4299d83d4295da9f498ac@%3Cdev.parquet.apache.org%3E

wesm commented on CMakeLists.txt:
```diff
@@ -674,13 +674,15 @@ set(LIBPARQUET_SRCS
+  src/parquet/murmur3.cc
   src/parquet/parquet_constants.cpp
   src/parquet/parquet_types.cpp
   src/parquet/printer.cc
```
What is this binary data file and can it be generated rather than checked into the repo?

cjjnjust replied:
The binary file is used to ensure cross compatibility between parquet-mr and parquet-cpp.

wesm replied:
I think we should store binary files someplace else, like in a parquet-compatibility repo. I can request ASF Infra to create such a repository if you think that's a good idea

cjjnjust replied:
Thanks, I agree that's a good idea. Other parquet implementation can benefit from that as well.  How long it will take to create such a repo and any dependency? I have no experience on this.

wesm commented on src/parquet/README:
```diff
@@ -0,0 +1,10 @@
+Step 1: Construct a Bloom filter with 1024 bytes of bitset.
+Step 2: Insert hashes of "hello", "parquet", "bloom", "filter" strings to Bloom filter
+by calling hash and insert APIs.
+Step 3: Call writeTo API to write to File.
```
I am not sure this is the right place for this. Put this comment in bloom_filter-test.cc?

cjjnjust replied:
We already have this doc in bloom_filter-test.cc.

wesm commented on src/parquet/bloom_filter.h:
```diff
@@ -0,0 +1,241 @@
+// A Bloom filter is a compact structure to indicate whether an item is not in a set or
+// probably in a set. The Bloom filter usually consists of a bit set that represents a
+// set of elements, a hash strategy and a Bloom filter algorithm.
+class BloomFilter {
```
Pretty much all of our hashing and general algorithmic code for parquet-cpp is contained in the `arrow/util` directory in Apache Arrow. Maybe this should be contributed there also?

cjjnjust replied:
We can open a jira to track this when we complete this, is that OK?

wesm commented on src/parquet/bloom_filter.cc:
```diff
@@ -0,0 +1,169 @@
+  }
+
+  num_bytes_ = num_bytes;
+  PARQUET_THROW_NOT_OK(::arrow::AllocateBuffer(pool_, num_bytes_, &data_));
```
I'm -1 on non-trivial constructors that can throw exceptions. Ideally a static alternate constructor returning `arrow::Status`

cjjnjust replied:
Maybe we can add a Init function to take over actual initialization work.

wesm commented on src/parquet/bloom_filter.cc:
```diff
@@ -0,0 +1,169 @@
+      hash_strategy_(HashStrategy::MURMUR3_X64_128),
+      algorithm_(Algorithm::BLOCK) {
+  if (!bitset) {
+    throw ParquetException("Given bitset is NULL");
```
Is this a pre-condition? If so use DCHECK

wesm commented on src/parquet/bloom_filter.cc:
```diff
@@ -0,0 +1,169 @@
+
+void BlockSplitBloomFilter::WriteTo(OutputStream* sink) const {
+  if (!sink) {
+    throw ParquetException("Given output stream is NULL.");
```
Probably should be DCHECK

wesm commented on src/parquet/bloom_filter.h:
```diff
@@ -0,0 +1,241 @@
+ public:
+  // Maximum Bloom filter size, it sets to HDFS default block size 128MB
+  // This value will be reconsidered when implementing Bloom filter producer.
+  static constexpr uint32_t MAXIMUM_BLOOM_FILTER_BYTES = 128 * 1024 * 1024;
```
Use `kMaximumBloomFilterBytes` naming style for better consistency with Google style guide

wesm commented on src/parquet/bloom_filter.h:
```diff
@@ -0,0 +1,241 @@
+  /// @param hash the element to contain.
+  /// @return false if value is definitely not in set, and true means PROBABLY
+  /// in set.
+  virtual bool FindHash(uint64_t hash) const = 0;
```
Having all of these virtual functions potentially on the hot path is concerning. What is the rationale for using virtual?

wesm commented on src/parquet/bloom_filter.h:
```diff
@@ -0,0 +1,241 @@
+//
+// This implementation sets 8 bits in each tiny Bloom filter. Each tiny Bloom
+// filter is 32 bytes to take advantage of 32-byte SIMD instructions.
+class BlockSplitBloomFilter : public BloomFilter {
```
Per my comment above, if there is only a single initial implementation of `BloomFilter`, putting virtual paths on the hot path seems potentially concerning. There are a number of approaches possible (like CRTP) to avoid virtual while permitting different implementations

wesm commented on src/parquet/bloom_filter.h:
```diff
@@ -0,0 +1,241 @@
+  BlockSplitBloomFilter(const uint8_t* bitset, uint32_t num_bytes);
+
+  // Minimum Bloom filter size, it sets to 32 bytes to fit a tiny Bloom filter.
+  static constexpr uint32_t MINIMUM_BLOOM_FILTER_BYTES = 32;
```
See comment re: name style for constants

wesm commented on src/parquet/hasher.h:
```diff
@@ -0,0 +1,75 @@
+  ///
+  /// @param value the value to hash.
+  /// @return hash result.
+  virtual uint64_t Hash(int32_t value) const = 0;
```
Is there an approach possible that does not use virtual?

jbapple-cloudera commented:
@wesm In answer to your question: yes, there are ways to address this without virtual. I originally suggested it as a way to address the open universe construct of knowing that we have a single filter API and multiple possible BF implementations and hash functions, even though we're starting with just one of each.

cjjnjust commented:
Thanks @wesm and @jbapple-cloudera , I saw the ARROW-2998 already committed.

The latest commit cannot be applied directly since parquet-cpp is using arrow repo at Dec, 2017 and it should have many changes from then on. I can create a jira to track ```shared_ptr ``` to ```unique_ptr``` under Bloom filter umbrella jira, is that OK?

jbapple-cloudera commented:
Yes, @cjjnjust , it's fine with me if we leave `shared_ptr` for now.

I think the largest unsettled issue is the discussion @wesm started about the performance of virtual methods.

Wes, I'd like to enable an API where users can write code something like

```
void ProcessFilter(const char * serialized_filter) {
  unique_ptr<Filter> f = Filter::Deserialize(serialized_filter);
  // process here
}
```

We could accomplish something similar with templates, but all of the processing and storage would then need to lift the type to a template parameter. I think we'll get a better picture of how onerous this is once we have some non-test use of the filter in parquet-cpp. How would you feel about revisiting that question then?

wesm commented:
I'm cool with tabling the virtual function discussion. I will be able to assist with refactoring for CRTP-ifying things when the time comes. I'd prefer to get the outstanding patches merged quickly and move as expediently as possible toward the Arrow-Parquet monorepo structure under discussion to make these things easier to work on

jbapple-cloudera commented:
OK, LGTM. Wes, since you're a committer, do you want to check this in and then we can work on some follow-on patches, like templatizing, integration with the rest of the codebase, a parquet-binary repo, and so on?

wesm commented:
+1. Merging this, sounds good. I created https://github.com/apache/parquet-testing; @cjjnjust let me know when you want to move the binary files there
